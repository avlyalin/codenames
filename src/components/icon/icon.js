import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

const Icon = React.forwardRef(function Icon(props, ref) {
  const {
    classes = '',
    color = 'default',
    filled = false,
    icon,
    text = '',
  } = props;
  const content = icon ? (
    <FontAwesomeIcon icon={icon} />
  ) : (
    <span className="text-sm uppercase font-bold">{text}</span>
  );

  let colorClasses = 'text-black bg-gray-200';
  if (color === 'blue') {
    if (filled) {
      colorClasses = 'bg-blue-100 text-white';
    } else {
      colorClasses = 'text-blue-100 bg-white';
    }
  } else if (color === 'red') {
    if (filled) {
      colorClasses = 'bg-red-100 text-white';
    } else {
      colorClasses = 'text-red-100 bg-white';
    }
  }

  return (
    <span
      ref={ref}
      className={classnames(
        'inline-flex justify-center items-center',
        'rounded-full',
        'h-8 w-8',
        colorClasses,
        classes,
      )}
    >
      {content}
    </span>
  );
});

Icon.propTypes = {
  classes: PropTypes.string,
  color: PropTypes.oneOf(['default', 'blue', 'red']),
  icon: PropTypes.string,
  text: PropTypes.string,
  filled: PropTypes.bool,
};

export { Icon };
