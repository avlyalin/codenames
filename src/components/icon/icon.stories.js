import React from 'react';
import { withKnobs, text, radios, boolean } from '@storybook/addon-knobs';
import { containerDecorator } from '_storybook/container';
import { Icon } from './icon';

// eslint-disable-next-line import/no-default-export
export default {
  title: 'Components/Icon',
  decorators: [containerDecorator(), withKnobs],
};

export const WithIcon = () => {
  const filled = boolean('Filled', false);
  const icon = text('Icon name', 'user-plus');
  const textContent = text('Text content', '');
  const color = radios(
    'Color',
    { default: 'default', blue: 'blue', red: 'red' },
    'default',
  );
  return <Icon color={color} icon={icon} text={textContent} filled={filled} />;
};

export const WithText = () => {
  const filled = boolean('Filled', false);
  const icon = text('Icon name', '');
  const textContent = text('Text content', 'vs');
  const color = radios(
    'Color',
    { default: 'default', blue: 'blue', red: 'red' },
    'default',
  );
  return <Icon color={color} icon={icon} text={textContent} filled={filled} />;
};
