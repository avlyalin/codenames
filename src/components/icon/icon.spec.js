import React from 'react';
import { render } from '@testing-library/react';
import { Icon } from 'src/components/icon/icon';

describe('<Icon />', () => {
  test('should render icon', () => {
    const { container } = render(<Icon icon="user-plus" />);
    expect(container.firstChild).toMatchSnapshot();
  });
  test('should render text', () => {
    const { container } = render(<Icon text="VS" />);
    expect(container.firstChild).toMatchSnapshot();
  });
  test('icon prop has higher priority than text', () => {
    const { container } = render(<Icon icon="crown" text="VS" />);
    expect(container.firstChild).toMatchSnapshot();
  });
  test('should set blue text color class', () => {
    const { container } = render(<Icon color="blue" icon="crown" />);
    expect(container.firstChild).toMatchSnapshot();
  });
  test('should have filled blue classes', () => {
    const { container } = render(
      <Icon color="blue" icon="crown" filled={true} />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
