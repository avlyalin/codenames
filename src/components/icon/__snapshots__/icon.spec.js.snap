// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Icon /> icon prop has higher priority than text 1`] = `
<span
  class="inline-flex justify-center items-center rounded-full h-8 w-8 text-black bg-gray-200"
>
  <svg
    aria-hidden="true"
    class="svg-inline--fa fa-crown fa-w-20 "
    data-icon="crown"
    data-prefix="fas"
    focusable="false"
    role="img"
    viewBox="0 0 640 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M528 448H112c-8.8 0-16 7.2-16 16v32c0 8.8 7.2 16 16 16h416c8.8 0 16-7.2 16-16v-32c0-8.8-7.2-16-16-16zm64-320c-26.5 0-48 21.5-48 48 0 7.1 1.6 13.7 4.4 19.8L476 239.2c-15.4 9.2-35.3 4-44.2-11.6L350.3 85C361 76.2 368 63 368 48c0-26.5-21.5-48-48-48s-48 21.5-48 48c0 15 7 28.2 17.7 37l-81.5 142.6c-8.9 15.6-28.9 20.8-44.2 11.6l-72.3-43.4c2.7-6 4.4-12.7 4.4-19.8 0-26.5-21.5-48-48-48S0 149.5 0 176s21.5 48 48 48c2.6 0 5.2-.4 7.7-.8L128 416h384l72.3-192.8c2.5.4 5.1.8 7.7.8 26.5 0 48-21.5 48-48s-21.5-48-48-48z"
      fill="currentColor"
    />
  </svg>
</span>
`;

exports[`<Icon /> should have filled blue classes 1`] = `
<span
  class="inline-flex justify-center items-center rounded-full h-8 w-8 bg-blue-100 text-white"
>
  <svg
    aria-hidden="true"
    class="svg-inline--fa fa-crown fa-w-20 "
    data-icon="crown"
    data-prefix="fas"
    focusable="false"
    role="img"
    viewBox="0 0 640 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M528 448H112c-8.8 0-16 7.2-16 16v32c0 8.8 7.2 16 16 16h416c8.8 0 16-7.2 16-16v-32c0-8.8-7.2-16-16-16zm64-320c-26.5 0-48 21.5-48 48 0 7.1 1.6 13.7 4.4 19.8L476 239.2c-15.4 9.2-35.3 4-44.2-11.6L350.3 85C361 76.2 368 63 368 48c0-26.5-21.5-48-48-48s-48 21.5-48 48c0 15 7 28.2 17.7 37l-81.5 142.6c-8.9 15.6-28.9 20.8-44.2 11.6l-72.3-43.4c2.7-6 4.4-12.7 4.4-19.8 0-26.5-21.5-48-48-48S0 149.5 0 176s21.5 48 48 48c2.6 0 5.2-.4 7.7-.8L128 416h384l72.3-192.8c2.5.4 5.1.8 7.7.8 26.5 0 48-21.5 48-48s-21.5-48-48-48z"
      fill="currentColor"
    />
  </svg>
</span>
`;

exports[`<Icon /> should render icon 1`] = `
<span
  class="inline-flex justify-center items-center rounded-full h-8 w-8 text-black bg-gray-200"
>
  <svg
    aria-hidden="true"
    class="svg-inline--fa fa-user-plus fa-w-20 "
    data-icon="user-plus"
    data-prefix="fas"
    focusable="false"
    role="img"
    viewBox="0 0 640 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M624 208h-64v-64c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v64h-64c-8.8 0-16 7.2-16 16v32c0 8.8 7.2 16 16 16h64v64c0 8.8 7.2 16 16 16h32c8.8 0 16-7.2 16-16v-64h64c8.8 0 16-7.2 16-16v-32c0-8.8-7.2-16-16-16zm-400 48c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
      fill="currentColor"
    />
  </svg>
</span>
`;

exports[`<Icon /> should render text 1`] = `
<span
  class="inline-flex justify-center items-center rounded-full h-8 w-8 text-black bg-gray-200"
>
  <span
    class="text-sm uppercase font-bold"
  >
    VS
  </span>
</span>
`;

exports[`<Icon /> should set blue text color class 1`] = `
<span
  class="inline-flex justify-center items-center rounded-full h-8 w-8 text-blue-100 bg-white"
>
  <svg
    aria-hidden="true"
    class="svg-inline--fa fa-crown fa-w-20 "
    data-icon="crown"
    data-prefix="fas"
    focusable="false"
    role="img"
    viewBox="0 0 640 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M528 448H112c-8.8 0-16 7.2-16 16v32c0 8.8 7.2 16 16 16h416c8.8 0 16-7.2 16-16v-32c0-8.8-7.2-16-16-16zm64-320c-26.5 0-48 21.5-48 48 0 7.1 1.6 13.7 4.4 19.8L476 239.2c-15.4 9.2-35.3 4-44.2-11.6L350.3 85C361 76.2 368 63 368 48c0-26.5-21.5-48-48-48s-48 21.5-48 48c0 15 7 28.2 17.7 37l-81.5 142.6c-8.9 15.6-28.9 20.8-44.2 11.6l-72.3-43.4c2.7-6 4.4-12.7 4.4-19.8 0-26.5-21.5-48-48-48S0 149.5 0 176s21.5 48 48 48c2.6 0 5.2-.4 7.7-.8L128 416h384l72.3-192.8c2.5.4 5.1.8 7.7.8 26.5 0 48-21.5 48-48s-21.5-48-48-48z"
      fill="currentColor"
    />
  </svg>
</span>
`;
